plugins/mongo_db_plugin/mongo_db_plugin.cpp

```diff
@@ -204,495 +310,527 @@ void mongo_db_plugin_impl::consume_blocks() {
+
+      try {
+         const auto& trx_header_value = bsoncxx::from_json( trx_header_json );
+         trans_doc.append( kvp( "transaction_header", trx_header_value ));
+         if( !signing_keys_json.empty()) {
+            const auto& keys_value = bsoncxx::from_json( signing_keys_json );
+            trans_doc.append( kvp( "signing_keys", keys_value ));
+         }
+      } catch( std::exception& e ) {
+         elog( "Unable to convert transaction JSON to MongoDB JSON: ${e}", ("e", e.what()));
+         elog( "  JSON: ${j}", ("j", trx_header_json));
+         elog( "  JSON: ${j}", ("j", signing_keys_json));
       }
    }
 
-   auto now = std::chrono::duration_cast<std::chrono::milliseconds>(
-         std::chrono::microseconds{fc::time_point::now().time_since_epoch().count()});
-
-   block_doc.append(kvp("block_num", b_int32{static_cast<int32_t>(block_num)}),
-                    kvp("block_id", block_id_str),
-                    kvp("prev_block_id", prev_block_id_str),
-                    kvp("timestamp", b_date{std::chrono::milliseconds{
-                          std::chrono::seconds{block.timestamp.operator fc::time_point().sec_since_epoch()}}}),
-                    kvp("transaction_merkle_root", block.transaction_mroot.str()),
-                    kvp("producer_account_id", block.producer.to_string()),
-                    kvp("pending", b_bool{true}));
-   block_doc.append(kvp("createdAt", b_date{now}));
-
-   if (!blocks.insert_one(block_doc.view())) {
-      elog("Failed to insert block ${bid}", ("bid", block_id));
+   if( !trx.actions.empty()) {
+      bsoncxx::builder::basic::array action_array;
+      for( const auto& act : trx.actions ) {
+         process_action( trx_id_str, act, action_array );
+      }
+      trans_doc.append( kvp( "actions", action_array ));
    }
 
-   int32_t msg_num = -1;
-   bool actions_to_write = false;
-   auto process_action = [&](const std::string& trans_id_str, mongocxx::bulk_write& bulk_msgs, const chain::action& msg) -> auto {
-      auto msg_oid = bsoncxx::oid{};
-      auto msg_doc = bsoncxx::builder::basic::document{};
-      msg_doc.append(kvp("_id", b_oid{msg_oid}),
-                     kvp("action_id", b_int32{msg_num}),
-                     kvp("transaction_id", trans_id_str));
-      msg_doc.append(kvp("authorization", [&msg](bsoncxx::builder::basic::sub_array subarr) {
-         for (const auto& auth : msg.authorization) {
-            subarr.append([&auth](bsoncxx::builder::basic::sub_document subdoc) {
-               subdoc.append(kvp("actor", auth.actor.to_string()),
-                             kvp("permission", auth.permission.to_string()));
-            });
+   if( start_block_reached ) {
+      act_num = 0;
+      if( !trx.context_free_actions.empty()) {
+         bsoncxx::builder::basic::array action_array;
+         for( const auto& cfa : trx.context_free_actions ) {
+            process_action( trx_id_str, cfa, action_array );
```

Comment:
context_free_actions need to marked with something to distinguish them from non-contrext-free-actions (actions).